src/v/cluster/service.cc

```diff
@@ -95,6 +95,17 @@ service::join_node(join_node_request&& req, rpc::streaming_context&) {
           join_node_reply{false, model::node_id{-1}});
     }
 
+    if (req.logical_version > expect_version) {
+        vlog(
+          clusterlog.warn,
+          "Rejecting join request from node {}, logical version {} > {}",
+          req.node,
```

Comment:
the interesting cases here would be

1. a rolling upgrade: a newer upgraded node wouldn't be "joining" in this case (i was just thinking that the same api were used for new nodes joining and existing nodes just showing up to participate).
2. we wouldn't be able to perform a rolling upgrade into a larger cluster (ie upgrade+node-add simultaneously)? not that we'd want to anyway...

? that's all i can think of

Reply:
>existing nodes just showing up to participate

Right, the API is used to register members of an existing cluster, but we'd only fast forward the feature table if the joining node were completely empty. For the typical upgrade path, wouldn't we still have data and join whatever version existed before the upgrade?

Reply:
>For the typical upgrade path, wouldn't we still have data and join whatever version existed before the upgrade?

i'm not sure this is always the case. for local-ssd k8s in cloud i'm pretty sure we nuke all the local data, upgrade as a new, empty node that joins the cluster.